File: LeetCode/Problem 5.py
# LeetCode: Problem 5
# Jonathan Kosasih

# Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.

# Example 1:

# Input: "babad"
# Output: "bab"
# Note: "aba" is also a valid answer.
# Example 2:

# Input: "cbbd"
# Output: "bb"
#------------------------------------------------------------------
class Solution(object):
    # initial solution
    def longestPalindrome(self, string):
        """
        :type s: str
        :rtype: str
        """
        array = []
        palindromes = []
        for char in string:
            for entry in range(len(array)):
                array[entry] += char
                if(array[entry] == array[entry][::-1]):
                    palindromes.append(array[entry])
            array.append(char)
        return print(max(palindromes))
    
    # dynamic programming
    def longestPalindromeDP(self, s):
        ans = ''
        max_len = 0
        n = len(s)
        DP = [[0] * n for _ in range(n)]
        for i in range(n):
            DP[i][i] = True
            max_len = 1
            ans = s[i]
        for i in range(n-1):
            if(s[i] == s[i+1]):
                DP[i][i+1] = True
                ans = s[i:i+2]
                max_len = 2
        for j in range(n):
            for i in range(0, j-1):
                if(s[i] == s[j] and DP[i+1][j-1]):
                    DP[i][j] = True
                    if(max_len < j - i + 1):
                        ans = s[i:j+1]
                        max_len = j - i + 1
        return ans
        
sol = Solution()
example = "babad"
#example = "cbbd"
#sol.longestPalindrome(example)
print(sol.longestPalindromeDP(example))
print("Test complete")

File: LeetCode/C++/Problem 5.cpp
// LeetCode: Problem 5
// Jonathan Kosasih
/*
Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.

Example 1:
Input: "babad"
Output: "bab"
Note: "aba" is also a valid answer.

Example 2:
Input: "cbbd"
Output: "bb"
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;

class Solution {
public:
	string longestPalindrome(string s) {
		if (s.empty()) {
			return "";
		}
		string ans = "";
		int maxLen = 0;
		int inputLength = s.size();
		bool dp[inputLength][inputLength] = { {false} };
		// initialize 1 letter palindromes
		for (int i = 0; i < inputLength; i++) {
			dp[i][i] = true;
		}
		maxLen = 1;
		ans = s[0];
		// initialize 2 letter palindromes
		for (int i = 0; i < inputLength-1 ; i++) {
			if (s[i] == s[i + 1]) {
				dp[i][i + 1] = true;
				ans = s.substr(i, 2);
				maxLen = 2;
			}
		}
		// fill in the rest with DP
		// go by column first
		// check with dp[i][j] = dp[i+1][j-1] and s[i] == s[j]
		for (int j = 2; j < inputLength; j++) {
			for (int i = 0; i < j; i++) {
				if (s[i] == s[j] && dp[i + 1][j - 1]) {
					dp[i][j] = true;
					if (maxLen < (j - i + 1)) {
						maxLen = j - i + 1;
						ans = s.substr(i, maxLen);
					}
				}
			}
		}
		return ans;
	}
};

int main()
{
    //std::cout << "Hello World!\n";
}